### PR TITLE
[RLlib] Issue 21334: Fix APPO when kl_loss is enabled.

### LINKED ISSUE #21334: [RLlib] [Bug] APPO with kl_loss learner_stats bug
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.


### Ray Component

RLlib

### What happened + What you expected to happen

Running an APPO training using `use_kl_loss: True` produces some errors.
Running a modified version of the APPO pendulum tuned example (by adding `use_kl_loss: True`):
```
rllib train -f rllib/tuned_examples/ppo/pendulum-appo.yaml
```

throws:
<details>
<summary>Stacktrace</summary>

```
ray.exceptions.RayTaskError(KeyError): ray::APPO.train_buffered() (pid=3768613, ip=192.168.1.74, repr=APPO)
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/tune/trainable.py", line 255, in train_buffered
    result = self.train()                                                                                
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/tune/trainable.py", line 314, in train
    result = self.step()                                                                                                                                                                                           
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/agents/trainer.py", line 885, in step
    raise e                                                                                                                                                                                                        
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/agents/trainer.py", line 867, in step
    result = self.step_attempt()                                                                                                                                                                                   
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/agents/trainer.py", line 920, in step_attempt
    step_results = next(self.train_exec_impl)                                                                                                                                                                      
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/util/iter.py", line 756, in __next__
    return next(self.built_iterator)                                                                     
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/util/iter.py", line 843, in apply_filter
    for item in it:
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/util/iter.py", line 843, in apply_filter
    for item in it:
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/util/iter.py", line 791, in apply_foreach
    result = fn(item)
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/agents/ppo/appo.py", line 107, in __call__
    self.update_kl(fetches)
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/agents/ppo/ppo.py", line 214, in __call__
    self.workers.local_worker().foreach_trainable_policy(update)
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/evaluation/rollout_worker.py", line 1256, in foreach_trainable_policy
    return [
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/evaluation/rollout_worker.py", line 1257, in <listcomp>
    func(policy, pid, **kwargs)
  File "/home/vince/.pyenv/versions/3.9.7/envs/spg-exp/lib/python3.9/site-packages/ray/rllib/agents/ppo/ppo.py", line 205, in update
    kl = fetches[pi_id][LEARNER_STATS_KEY].get("kl") 
KeyError: 'learner_stats'
```
</details>



My code produces the stack trace below. I tried to make a minimal reproduction, but then I run into the error mentioned above, so I stopped (but I can dig into this if needed). My code uses Tune to run the experiments, maybe that causes the difference. I added the print statements to `rllib/utils/metrics/learner_info.py` to see what triggers the crash. 
I think this is related to the issue above, but I'm not sure.

<details>
<summary>Stacktrace</summary>

```
(APPO pid=3782547) #########                                                                                                                                                                                       
(APPO pid=3782547) ('learner_stats', 'KL_Coeff')                                                                                                                                                                   
(APPO pid=3782547) #########                                                                                                                                                                                       
(APPO pid=3782547) ('learner_stats', 'cur_lr')                                                                                                                                                                     
(APPO pid=3782547) #########                                                                                                                                                                                       
(APPO pid=3782547) ('learner_stats', 'entropy')                                                                                                                                                                    
(APPO pid=3782547) #########                                                                                                                                                                                       
(APPO pid=3782547) ('learner_stats', 'entropy_coeff')                                                                                                                                                              
(APPO pid=3782547) #########                                                                                                                                                                                       
(APPO pid=3782547) ('learner_stats', 'kl', 0)                                                                                                                                                                      
(APPO pid=3782547) Exception in thread Thread-1:                                                                                                                                                                   
(APPO pid=3782547) Traceback (most recent call last):                                                                                                                                                              
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/3.9.7/lib/python3.9/threading.py", line 973, in _bootstrap_inner
(APPO pid=3782547)     self.run()
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/ray/rllib/execution/learner_thread.py", line 69, in run
(APPO pid=3782547)     self.step()
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/ray/rllib/execution/multi_gpu_learner_thread.py", line 162, in step
(APPO pid=3782547)     learner_info_builder.add_learn_on_batch_results(
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/ray/rllib/utils/metrics/learner_info.py", line 44, in add_learn_on_batch_results
(APPO pid=3782547)     tree.map_structure_with_path(
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/tree/__init__.py", line 549, in map_structure_with_path
(APPO pid=3782547)     return map_structure_with_path_up_to(structures[0], func, *structures,
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/tree/__init__.py", line 852, in map_structure_with_path_up_to
(APPO pid=3782547)     [func(*args) for args in zip(flat_path_list, *flat_value_lists)])
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/tree/__init__.py", line 852, in <listcomp>
(APPO pid=3782547)     [func(*args) for args in zip(flat_path_list, *flat_value_lists)])
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/ray/rllib/utils/metrics/learner_info.py", line 45, in <lambda>
(APPO pid=3782547)     lambda p, *s: all_tower_reduce(p, *s),
(APPO pid=3782547)   File "/home/vince/.pyenv/versions/spg-exp/lib/python3.9/site-packages/ray/rllib/utils/metrics/learner_info.py", line 80, in all_tower_reduce
(APPO pid=3782547)     if path[-1].startswith("min_"):
(APPO pid=3782547) AttributeError: 'int' object has no attribute 'startswith'
```
</details>


### Versions / Dependencies

Python: 3.9.7
Ray: 1.9.0

### Reproduction script

See above.

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

Fix APPO agent when kl_loss is enabled.
value is now saved under different policy id keys. also we need to torch_mean() the stats for the torch policy.

## Related issue number

Closes #21334

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/ppo/tests/test_appo.py [hunk 1]
```diff
@@ -47,6 +47,25 @@ def test_appo_compilation(self):
             check_compute_single_action(trainer)
             trainer.stop()
 
+    def test_appo_compilation_use_kl_loss(self):
+        """Test whether an APPOTrainer can be built with both frameworks."""
```
Comment (sven1977) [inline rllib/agents/ppo/tests/test_appo.py:51]: Nit: Fix the comment?

File: rllib/agents/ppo/tests/test_appo.py [hunk 2]
```diff
@@ -47,6 +47,25 @@ def test_appo_compilation(self):
...
+        config = ppo.appo.DEFAULT_CONFIG.copy()
+        config["num_workers"] = 1
+        config["use_kl_loss"] = True
+        num_iterations = 2
+
+        for _ in framework_iterator(config, with_eager_tracing=True):
+            print("w/ v-trace")
```
Comment (sven1977) [inline rllib/agents/ppo/tests/test_appo.py:58]: Not necessary here, no?

File: rllib/agents/ppo/tests/test_appo.py [hunk 3]
```diff
@@ -47,6 +47,25 @@ def test_appo_compilation(self):
...
+            _config = config.copy()
+            _config["vtrace"] = True
```
Comment (sven1977) [inline rllib/agents/ppo/tests/test_appo.py:60]: same

File: rllib/execution/multi_gpu_learner_thread.py [hunk 4]
```diff
@@ -166,15 +165,12 @@ def step(self) -> None:
...
-            learner_stats = {
-                pid: self.learner_info[pid][LEARNER_STATS_KEY]
-                for pid in self.learner_info.keys()
-            }
 
         if released:
             self.idle_tower_stacks.put(buffer_idx)
 
-        self.outqueue.put((get_num_samples_loaded_into_buffer, learner_stats))
+        self.outqueue.put((get_num_samples_loaded_into_buffer,
```
Comment (sven1977) [inline rllib/execution/multi_gpu_learner_thread.py:172]: Nice!

**Review by sven1977 — APPROVED** [review]
Very cool! Thanks for the fix.
Is there an issue related to this PR? Could you change the title to: `[RLlib] Issue xyz: ...`

File: rllib/agents/ppo/tests/test_appo.py (hunk 1 above)
Comment (gjoliver) [reply rllib/agents/ppo/tests/test_appo.py:51]: oops done.

File: rllib/agents/ppo/tests/test_appo.py (hunk 2 above)
Comment (gjoliver) [reply rllib/agents/ppo/tests/test_appo.py:58]: right, I got rid of it.

File: rllib/agents/ppo/tests/test_appo.py (hunk 3 above)
Comment (gjoliver) [reply rllib/agents/ppo/tests/test_appo.py:60]: done

**Pull request merged**